### DISCUSSION TITLE
Why is the performance of my custom functions so different from that of numerical operations?

### DISCUSSION THREAD

**Discussion by wodejiong** [trinodb/trino#23220] [Q&A] [answered]
I implemented a date_trunc operator that handles one of timestamps.
![image](https://github.com/user-attachments/assets/28d89d57-f873-4752-a521-9f80fb3f2f06)

It takes about 3 to 4 seconds to execute the following SQL statement:
SELECT date_trunc('hour', __time__) AS time, count(*)AS PV FROM log GROUP BY time ORDER BY time
![image](https://github.com/user-attachments/assets/c0ae3727-3e41-4337-bbcd-3644c29b3b37)

However, when I do numerical calculations, similar queries, with the same amount of data, are within 1 second.
SELECT (__time/(60*60*1000)*(60*60*1000)) AS time, count(*)AS PV FROM log GROUP BY time ORDER BY time
![image](https://github.com/user-attachments/assets/65014f34-9497-4974-acca-8c7b41c14efc)

What's the difference between the two, and how do I optimize my operator.

**Comment by hashhar** [comment, marked as answer]
The logic is obviously much more involved in the `date_trunc` case compared to just numerical operations. You should consider recording a JFR profile to see where the time is spent in both cases.

Also for some operations IIRC the engine can apply them in parallel to an entire `Page` instead of per-row (I'm not 100% sure about this to be honest).
